norse/torch/module/snn.py

```diff
@@ -0,0 +1,201 @@
+            tensor, and parameters module, and returning a tuple of (output spikes, state).
+        state_fallback (Callable[[torch.Tensor], torch.Tensor]): A function that can return a
+            default state with the correct dimensions, in case no state is provided in the
+            forward pass.
+        p (torch.nn.Module): The neuron parameters as a torch Module, which allows the module
+            to configure neuron parameters as optimizable.
+        dt (float): Time step to use in integration. Defaults to 0.001.
+    """
+
+    def __init__(
+        self,
+        activation: FeedforwardActivation,
+        state_fallback: Callable[[torch.Tensor], torch.Tensor],
+        p: torch.nn.Module,
+        dt: float = 0.001,
+    ):
+        super().__init__()
+        self.activation = activation
+        self.state_fallback = state_fallback
+        self.p = p
+        self.dt = dt
+
+    def forward(self, input_tensor: torch.Tensor, state: Optional[torch.Tensor]):
+        state = state if state is not None else self.state_fallback(input_tensor)
+        return self.activation(input_tensor, state, self.p, self.dt)
+
+
+class RecurrentSNNCell(torch.nn.Module):
+    """
+    The base module for recurrent spiking neural networks (RSNN) *without* time.
+
+    Parameters:
+        input_size (int): The number of input neurons
+        hidden_size (int): The number of hidden neurons
+        activation (RecurrentActivation): The activation function accepting an input tensor, state
+            tensor, input weights, recurrent weights, and parameters module, and returning a tuple
+            of (output spikes (one per timestep), state).
+        state_fallback (Callable[[torch.Tensor], torch.Tensor]): A function that can return a
+            default state with the correct dimensions, in case no state is provided in the
+            forward pass.
+        p (torch.nn.Module): The neuron parameters as a torch Module, which allows the module
+            to configure neuron parameters as optimizable.
+        input_weights (torch.Tensor): Weights used for input tensors. Defaults to a random
+            matrix normalized to the number of hidden neurons.
+        recurrent_weights (torch.Tensor): Weights used for input tensors. Defaults to a random
+            matrix normalized to the number of hidden neurons.
+        autopses (bool): Allow self-connections in the recurrence? Defaults to False. Will also
+            remove autopses in custom recurrent weights, if set above.
+        dt (float): Time step to use in integration. Defaults to 0.001.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        activation: FeedforwardActivation,
+        state_fallback: Callable[[torch.Tensor], torch.Tensor],
+        p: torch.nn.Module,
+        input_weights: Optional[torch.Tensor] = None,
+        recurrent_weights: Optional[torch.Tensor] = None,
+        autopses: bool = False,
```

Comment:
Shouldn't this be autapse?

Reply:
Don't know. We also write weight**s**. Theoretically, it is about killing *all* the autopse**s** between all the neurons in the module. But would it be easier to understand in singular?

Reply:
Ah, you meant the spelling :laughing:! Yes, that completely true! But would you also do singular?

Reply:
No plural is fine, I just meant the 'a' :smile: